### PR TITLE
net: tcp: Improve thread safety of the TCP stack

### PR DESCRIPTION
**Pull request by rlubos**
Improve thread safety of the TCP stack by eliminating race between TCP input thread and TCP workqueue. Additionally, fix the thread-safety issue with the connection module. Please see the details in the commit messages.

Those race conditions are rather hard to trigger in a regular environment, but a malicious device could easily perform an attack by flooding the network with crafted packets. The test scenario here was as follows:
* TCP client just establishes a TCP connection with the server, and waits for the connection to close (reset)
* A malfunctioning TCP server, accepts the connection and then floods the client with the stream of RST packets. The first one triggers the client to close the connection, while still having to process the remaining incoming RST duplicates.

W/o those fixes, this ended with an almost immediate crash. With the TCP fix only, it usually took a few hundreds of connection attempts to crash (it is quite random). With both fixes included, no crash was observed yet.

### CODE REVIEW

**Review by jukkar — DISMISSED** [review]

**Review by pdgendt — DISMISSED** [review]

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -2428,6 +2437,11 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 
 	k_mutex_lock(&conn->lock, K_FOREVER);
 
+	/* Connection context was already freed. */
+	if (conn->state == TCP_UNUSED) {
+		return NET_DROP;
```
Comment (talih0) [inline subsys/net/ip/tcp.c:2442]: nit: you may want to add to the comment that a k_mutex_unlock() is not needed before return NET_DROP. (I'm assuming it's not)

Also, we are relying that nothing else touches this memory after k_mem_slab_free(). Can anything call tcp_conn_alloc() and use this memory in some other race condition?

**Review by talih0 — COMMENTED** [review]
Left comment in the diff.

**Review by jukkar — APPROVED** [review]

File: subsys/net/ip/tcp.c (hunk 1 above)
Comment (talih0) [reply subsys/net/ip/tcp.c:2443]: > 
> In any case however, do you have any idea how could we protect against such case? I can't really think of any good countermeasure other than a complete redesign on how connection contexts are allocated (i. e. connection-related mutexes should only be initialized once, on boot).

I can't see any simple ways either.

**Review by talih0 — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by yashi** [comment]
Just want to let every know that this issue can easily be triggered by asscessing a coap server on Zephyr from two or more client using [aiocoap](https://aiocoap.readthedocs.io/) from a PC.

**Pull request merged**